File: Labs/Lab1 CG/Lab1 CG/Deck.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab1_CG
{
    internal class Deck
    {
        private List<Card> cards;
        private Random random = new Random();

        public Deck()
        {
            this.cards = new List<Card>();
        }

        public Deck(List<Card> c)
        {
            this.cards = new List<Card>();

            if (c != null)
            {
                foreach (Card card in c)
                {
                    this.cards.Add(card);
                }
            }
        }

        public void addCard(Card card)
        {
            if (card != null)
            {
                cards.Add(card);
            }
        }

        public bool removeCard(Card card)
        {
            return cards.Remove(card);
        }

        public void clear()
        {
            cards.Clear();
        }


        public override string ToString()
        {
            StringBuilder stringBuilder = new StringBuilder("Deck:\n");

            foreach (Card card in cards)
            {
                stringBuilder.Append(card + "\n");
            }

            return stringBuilder.ToString();
        }

        public Card getRandomCard()
        {
            if (cards.Count() == 0)
            {
                return null;
            }

            int randomIndex = random.Next(cards.Count() - 1);

            Card c = cards[randomIndex];
            cards.RemoveAt(randomIndex);

            return c;
        }

        public List<Card> Cards { get => cards.ToList(); }

        public Card topCard()
        {
            return cards[cards.Count - 1];
        }

        public Card drawCard()
        {
            Card c = cards[cards.Count - 1];
            cards.RemoveAt(cards.Count - 1);

            return c;
        }

        public void putCard(Card card)
        {
            cards.Add(card);
        }

        public void sortByValue()
        {
            cards.Sort();
        }

        public void shuffle()
        {
            int n = cards.Count;

            while (n > 1)
            {
                n--;
                int k = random.Next(n + 1);

                Card value = cards[k];
                cards[k] = cards[n];
                cards[n] = value;
            }
        }
    }
}

File: Labs/Lab1 CG/CardGamesJava/src/CardGames/Deck.java
package CardGames;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.Random;
import java.util.Stack;

/**
 * Esta classe representa um baralho de cartas
 *
 * @author POO 2019/2020
 * @version maio/2020
 */
public class Deck {

    private Stack<Card> cards;  
    private Random random = new Random();

    public Deck() {
        cards = new Stack<>(); 
    }

    public Deck(List<Card> cards) {
        this.cards = new Stack<>();
        if (cards != null) {
            for (Card card : cards) {
                this.cards.push(card);
            }
        }
    }

    public void addCard(Card card) {
        if (card != null) {
            cards.add(card);
        }
    }

    public boolean removeCard(Card card) {
        return cards.remove(card);
    }

    public void clear() {
        cards.clear();
    }

    @Override
    public String toString() {
        String cardsList = "";

        for (Card card : cards) {
            cardsList += card + "\n";
        }

        return cardsList;
    }

    public Card getRandomCard() {
        if (cards.isEmpty()) {
            return null;
        }
        int randomIndex = random.nextInt(cards.size() - 1);
        return cards.remove(randomIndex);
    }

    public void sortByValue() {
        Collections.sort(cards);
    }

    public Iterator<Card> iterator() {
        return cards.iterator();
    }

    public void shuffle() {
        Collections.shuffle(cards);
    }
    
    public List<Card> getCards() {
        return new ArrayList<Card>(cards);
    }

    public Card topCard() {
        return cards.peek();
    }

    public Card drawCard() {
        return cards.pop();
    }

    public Card putCard(Card card) {
        return cards.push(card);
    }
    
    public void sort() {
        Collections.sort(cards, new CardSuitValueComparator());
    }

   
}
